storage/storage.go

```diff
@@ -424,14 +440,57 @@ func (storage DBStorage) WriteReportForCluster(
 		return err
 	}
 
+	deleteQuery := "DELETE FROM rule WHERE org_id = $1 AND cluster_id = $2;"
+	_, err = tx.Exec(deleteQuery, orgID, clusterName)
+	if err != nil {
+		log.Err(err).Msgf("Unable to remove previous cluster reports (org: %v, cluster: %v)", orgID, clusterName)
+		_ = tx.Rollback()
+		return err
+	}
+
+	// Perform the report upsert.
+	reportedAtTime := time.Now()
+
+	for _, rule := range rules {
+		_, err = tx.Exec(ruleUpsertQuery, orgID, clusterName, rule.Module, rule.ErrorKey, string(rule.TemplateData))
+		if err != nil {
+			log.Err(err).Msgf("Unable to upsert the cluster report (org: %v, cluster: %v)", orgID, clusterName)
+			_ = tx.Rollback()
+			return err
+		}
+	}
+
+	_, err = tx.Exec(reportUpsertQuery, orgID, clusterName, report, reportedAtTime, lastCheckedTime, kafkaOffset)
+	if err != nil {
+		log.Err(err).Msgf("Unable to upsert the cluster report (org: %v, cluster: %v)", orgID, clusterName)
+		_ = tx.Rollback()
+		return err
+	}
+	return tx.Commit()
+}
+
+// WriteReportForCluster writes result (health status) for selected cluster for given organization
+func (storage DBStorage) WriteReportForCluster(
+	orgID types.OrgID,
+	clusterName types.ClusterName,
+	report types.ClusterReport,
+	rules []types.ReportItem,
+	lastCheckedTime time.Time,
+	kafkaOffset types.KafkaOffset,
+) error {
+	// Skip writing the report if it isn't newer than a report
+	// that is already in the database for the same cluster.
+	if oldLastChecked, exists := storage.clustersLastChecked[clusterName]; exists && !lastCheckedTime.After(oldLastChecked) {
+		return types.ErrOldReport
+	}
+
 	// Check if there is a more recent report for the cluster already in the database.
-	rows, err := tx.Query(
```

Comment:
don't removing this operation from transaction cause prohibition of atomic upsert operation?

Reply:
I didn't quite understand your question.

Reply:
I mean if the query is called outside transaction, there's no guarantee you can trust it's resultset later. Some other process can update database in meantime.

Reply:
hmm... ok, it makes sense